### ISSUE TITLE
multi-level command line parsing

### ISSUE THREAD

**Issue by maxpietsch** [MRtrix3/mrtrix3#1840]
As discussed here is a set of proposed command-line interface solutions that allow specifying grouped scalar or vector-valued options. 

The examples below are supposed to register 2 images `i` and `j` with 2 contrasts each (`w` and `g`).   The options can be specific to contrast (`-mc_weight`), registration stage (`-scale`), or both (`lmax`, for simplicity, I am ignoring the registration-type groupings (`rigid_lmax`, ..., `nl_max`)).

(1) colon separated arguments, repeated options (currently implemented solution in #985, replaced the plus with colon symbol):

```
mrregister iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1 -mc_weight 2,1 \
-lmax 0,2,4 -lmax 0,0,0 -transformed iw_t.mif jw_t.mif -transformed ig_t.mif jg_t.mif
```
(1b) drop the colon separator as not needed

```
mrregister iw.mif jw.mif ig.mif jg.mif -scale 0.3,0.5,1 -mc_weight 2,1 \
-lmax 0,2,4 -lmax 0,0,0 -transformed iw_t.mif jw_t.mif -transformed ig_t.mif jg_t.mif
```

For both, the order of repeated options defines which contrast it refers to. These have to be repeated once or for each contrast. Otherwise, the order does not matter, (1b) is equivalent to:

```
mrregister -scale 0.3,0.5,1 -mc_weight 2,1 \
iw.mif jw.mif -lmax 0,2,4 -transformed iw_t.mif jw_t.mif \
ig.mif jg.mif -lmax 0,0,0 -transformed ig_t.mif jg_t.mif
```

(2) colon-grouped parsing to separate contrasts:

```
mrregister iw.mif jw.mif -scale 0.3,0.5,1 -lmax 0,2,4 -transformed iw_t.mif jw_t.mif \
-mc_weight 2 : ig.mif jg.mif -lmax 0,0,0 -transformed ig_t.mif jg_t.mif -mc_weight 1
```

`-scale 0.3,0.5,1` can be specified at any location.

(3) colon-separated arguments or option arguments (as (1) but option arguments are grouped):

```
mrregister iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1 -lmax 0,2,3 : 0,0,0 \
-transformed iw_t.mif jw_t.mif : ig_t.mif jg_t.mif -mc_weight 2 : 1
```

----
(1) and (1b) requires very few changes to the existing registration code. 

The advantage of (2) is the clear visual separation of contrasts but it restricts the user to group arguments by contrast.

I have a slight preference for (3) if that is possible in the option parsing backend as options can be added without jumping around on the terminal assuming one starts with a fixed set of contrasts. 

Opinions?

**Comment by jdtournier** [comment]
Option (3) is not possible without some very custom handling. Essentially, it can only really work if you mandate that _all_ arguments must appear before group-able options, so that you can actually figure out how many arguments each option is supposed to accept. Currently, the number of arguments expected for each option needs to be fixed and known at compile-time. Otherwise it's impossible to tell its arguments apart from generic command-line arguments, especially if it occurs earlier than the last actual argument - e.g. for your example (stripping out the second line of it to keep it simple):
```
mrregister -lmax 0,2,3 : 0,0,0 iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1 
```
How would the code figure out in a simple robust way that `-lmax` expects 3 arguments...?

**Comment by jdtournier** [comment]
One potentially way around this is to allow the use of the colon as part of a single argument, e.g.:
```
mrregister iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1 -lmax 0,2,3:0,0,0 
```
or potentially
```
mrregister iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1 -lmax "0,2,3 : 0,0,0"
```

But it still doesn't work for the `-transformed` option, for the same reasons: you're expecting it to handle a variable number of arguments - and I can't see a way to pass those as single arguments in that case...

**Comment by maxpietsch** [comment]
> How would the code figure out in a simple robust way that -lmax expects 3 arguments...?

I might be missing the point but can't we just make colon do the same as repeated options in the option parsing backend. Then it's a matter of assigning colons to arguments or options and ignore them for arguments. For instance:

```python
args = 'mrregister -lmax 0,2,3 : 0,0,0 iw.mif jw.mif : ig.mif jg.mif -scale 0.3,0.5,1'.split()

refined_args = []
for i, arg in enumerate(args):
    if arg == ':':
        continue
    if i > 1 and args[i-1] == ':' and len(refined_args) > 1 and refined_args[-2].startswith('-'):
        refined_args[-1] = refined_args[-1] + ':' + arg
    else:
        refined_args.append(arg)
print(refined_args)
# ['mrregister', '-lmax', '0,2,3:0,0,0', 'iw.mif', 'jw.mif', 'ig.mif', 'jg.mif', '-scale', '0.3,0.5,1']
```

**Comment by jdtournier** [comment]
Yep, that can work. But like I said, that's some very custom handling...

**Comment by Lestropie** [comment]
I like how neither of you have cottoned on to the issue of "`-lmax 0,2,3`"... :-P

-----

I would propose (2a):

```
mrregister -scale 0.3,0.5,1 : \ 
iw.mif jw.mif -lmax 0,2,4 -transformed iw_t.mif jw_t.mif -mc_weight 2 : \
ig.mif jg.mif -lmax 0,0,0 -transformed ig_t.mif jg_t.mif -mc_weight 1
```

`-scale 0.3,0.5,1` must be specified before the first colon, as it is a global option that always applies to all contrasts, and thus must be explicitly separated from the per-contrast arguments & options. In this case, you would need an additional boolean flag for command-line options [here](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/core/cmdline_option.h#L64-L67), something like "IsArgGroupWise", which distinguishes between:

- Options that apply globally and hence *must* appear before the first colon (if any are present);

- Options that *cannot* appear before the first colon (if present) as they apply only to "argument groups".

**Comment by maxpietsch** [comment]
> Yep, that can work. But like I said, that's some very custom handling...

Yes, the question is, what is most useful long-term or intuitive for users and is it worth the effort to implement this in the backend if most (all commands but `mrregister`) do not require this.

(2) and (2a) group the command line, (3) groups on the option level.

(2a) is the clearest about to what options refer to. In favour of (3*) over (2*) is that it is likely more useful to provide a mechanism for matrix-valued inputs for options than to globally section the command line. 

Could the python argument parser handle (2*)? For (1), all I had to do was to add a [PlusSeparator](https://github.com/MRtrix3/mrtrix3/blob/d45c149380699d61f8f7531d4d1813a781720c68/lib/mrtrix3/app.py#L541-L556) class to the parser. For this to also apply to options (3*), modifying sys.args before argument parsing happens using the [above code](https://github.com/MRtrix3/mrtrix3/issues/1840#issuecomment-564173935) should suffice I think.

**Comment by maxpietsch** [comment]
There is one usage of the colon separator already in the library that would require special handling to disambiguate range and grouping operations (`-axis` can be repeated): 
`mrgrid in.mif crop -axis index <start>:<stop> out.mif`
`mrgrid in.mif crop -axis index <start>: out.mif`
`mrgrid in.mif crop -axis index <start>:end out.mif`
and 
`mrgrid in.mif crop -axis index <delta_lower>,<delta_upper> out.mif`

The range-based use currently requires a non-whitespace character before `:`.

**Comment by jdtournier** [comment]
Indeed, I wouldn't personally be too keen on using a colon in as a second-level separator in these options  (even though I've actually suggested it...), since it's the range operator for [numbered sequences](https://mrtrix.readthedocs.io/en/latest/getting_started/command_line.html#number-sequences-and-floating-point-lists). 

I think option (3) is going to require totally custom handling, specific to `mrregister`. Not impossible, but not massively desirable either if we can avoid it. 

Rob's option (2a) is the cleanest if you need to have a separator, I reckon. It cleanly separates the overall options from the options specific to each 'group'. From the code side, we would need to keep track of the arguments and options within each 'group', and provide functions to query the number of groups, and switch between group (defaulting to group 0, which would avoid the need to modify any other command). 

As an alternative solution, we could allow each option to keep track of the index of the preceding argument, and provide a method in the code to query that. This would mean that we could figure out which (pair of) argument(s) each option applies to in situations like this. We could then document (and enforce) that certain options apply to the previous argument, which would allow to write something like this:
```
mrregister -scale 0.3,0.5,1 \ 
iw.mif jw.mif -lmax 0,2,4 -transformed iw_t.mif jw_t.mif -mc_weight 2 \
ig.mif jg.mif -lmax 0,0,0 -transformed ig_t.mif jg_t.mif -mc_weight 1
```
which is basically (2a) _without_ the colon as a separator. In the code, you'd just need to figure out which contrast each e.g. `-lmax` option applies to by querying whether it appears after argument 2 or argument 4. You could also enforce that it can't appear after odd arguments, although that wouldn't be strictly required (1 or 2 would apply to the first contrast, 3 or 4 to the second) - but it would admittedly be odd for the user to write their command like that.

I reckon that last option is probably the least intrusive, and the functionality will likely prove useful in other situations too. Thoughts?

**Comment by maxpietsch** [comment]
When you compare 
```
mrregister -scale 0.3,0.5,1 \ 
iw.mif jw.mif -lmax 0,2,4 -transformed iw_t.mif jw_t.mif -mc_weight 2 \
ig.mif jg.mif -lmax 0,0,0 -transformed ig_t.mif jg_t.mif -mc_weight 1
```
to for instance

`dwifod dwi.mif rf_w wm.mif  rf_g gm.mif -lmax 4,0`
we'd be inconsistent in what options refer to and I think we need some visual queue for separating the logic of how we group options and arguments.

Let's assume we add two rules to the general command-line parsing:
1. remove leading and trailing whitespace around '::' before the argument parsing happens
2. arguments and options containing `::` (if flagged to allow this) are split and each part treated as a repeated option (similar for arguments).

Then
```
mrregister -lmax 0,2,5 :: 0,0,0 iw.mif :: ig.mif jw.mif :: jg.mif -mc_weight 2 :: 1 \
-transformed iw_t.mif :: ig_t.mif jw_t.mif :: jg_t.mif -scale 0.3,0.5,1
```
is equivalent to
```
mrregister -lmax 0,2,5::0,0,0 iw.mif::ig.mif jw.mif::jg.mif -mc_weight 2::1 \
-transformed iw_t.mif::ig_t.mif jw_t.mif::jg_t.mif -scale 0.3,0.5,1
```
Let's call this (3b).

With this, tab-complete works, we are explicit in what level each option refers to (all options or arguments without `::`-delimiter apply to all levels), it is shorter to type, and the order of options does not matter.

Side notes: 
- We can't use both multi-dimensional options and leave individual elements to the default if we repeat the stacking operator for both (`1:::2` could be [[[1]],[[2]]] or [1,default,default,2]).
- I've changed the argument logic from `iw.mif jw.mif :: ig.mif jg.mif` to `iw.mif :: ig.mif jw.mif :: jg.mif`

**Comment by Lestropie** [comment]
I personally don't  find "`iw.mif :: ig.mif jw.mif :: jg.mif`" intuitive. It's basically a local argument list grouping *a la* `[]`, but using an internal delimiter `::` rather than surrounding in brackets in order to permit tab-completion. But the absence of encapsulating symbols means that the grouping of items on the command-line is not immediately obvious; i.e.:

```
mrregister -lmax 0,2,4 :: 0,0,0 iw.mif :: ig.mif jw.mif :: jg.mif -mc_weight 2 :: 1 \
-transformed iw_t.mif :: ig_t.mif jw_t.mif :: jg_t.mif -scale 0.3,0.5,1
```

is actually grouped like:
```
mrregister
[-lmax 0,2,4 0,0,0]
[iw.mif ig.mif]
[jw.mif jg.mif]
[-mc_weight 2 1]
[-transformed iw_t.mif ig_t.mif]
[jw_t.mif jg_t.mif]
[-scale 0.3,0.5,1]
```
, but this grouping really isn't visually obvious from the usage.

In comparison:
```
mrregister -scale 0.3,0.5,1 :: iw.mif jw.mif -lmax 0,2,4 -mc_weight 2 -transformed iw_t.mif :: \
ig.mif jg.mif -lmax 0,0,0 -mc_weight 1 -transformed ig_t.mif
```
may have repeated options, but it has far fewer delimiters, and the grouping is therefore far more obvious:
```
mrregister
[-scale 0.3,0.5,1]
[iw.mif jw.mif -lmax 0,2,4 -mc_weight 2 -transformed iw_t.mif]
[ig.mif jg.mif -lmax 0,0,0 -mc_weight 1 -transformed ig_t.mif]
```

-----

The conflict with `dwi2fod` is there, but we should consider all options that are available in that respect. For instance, if we were to convert:
```
dwifod dwi.mif rf_w wm.mif rf_g gm.mif -lmax 4,0
```
to:
```
dwifod dwi.mif :: rf_w wm.mif -lmax 4 :: rf_g gm.mif -lmax 0
```
for consistency, that might not actually be so bad of a change.

Otherwise we would simply need to rely on the help description for the `-lmax` option depending on how it is interpreted by different commands. This is actually already the case: some commands accept a comma-delimited list to `-lmax`, others demand a single integer.

**Comment by dchristiaens** [comment]
Hi. Just my two cents on this discussion, after some light encouragement from @maxpietsch.

1. Personally, I do not think that syntax (in)consistency with `dwi2fod` is of much importance. The two commands serve completely different purposes, and I doubt that users would ever hold it against us. So, as far as I am concerned, `dwi2fod` can stay as it is.

2. I would _intuitively_ group input channels across contrasts, i.e., `[iw.mif ig.mif ...]` (sources) and `[jw.mif jg.mif ...]` (targets) as in suggestion (3a) above. However, the issue seems to be that all `mrregister` options either apply globally or to one individual contrast. In the absence of any "subject-wide" options, I can see why most of the possibilities put forward in this discussion group options per contrast.

3. Of the suggestions discussed so far, I like the `::` syntax best. My only concern is that additional separation between global options. Ideally, I would prefer the command syntax to remain unchanged for single-contrast registration.

**Comment by maxpietsch** [comment]
Candidate (4):

- `-option option_args` is for global options
- `--option option_args` is for contrast-related options, this is made clear in the command help page
- contrast-specific options are provided via repetition
- in the absence of multiple contrasts, `--option` can also be provided as `-option` 
- options need to be specified for one or all contrasts

```
mrregister -scale 0.3,0.5,1 \
 iw.mif jw.mif --mc_weight 2 --lmax 0,2,4 --transformed iw_t.mif jw_t.mif \
--contrast ig.mif jg.mif --mcweight 1 --lmax 0,0,0 --transformed ig_t.mif jg_t.mif
```
which is equivalent to
```
mrregister iw.mif jw.mif --contrast ig.mif jg.mif  \
--mc_weight 2 --mcweight 1 \
--lmax 0,2,4 --lmax 0,0,0 -scale 0.3,0.5,1 \
--transformed iw_t.mif jw_t.mif  --transformed ig_t.mif jg_t.mif
```

advantages:
- no delimiter needed
- no change to existing commands necessary
- maximally flexible in terms of where options appear
- explicit about what options refer to
- users can group their command line by contrast or by option with the same outcome

disadvantage:
- defaults can only apply to all contrasts

-----

If one is used to single contrast registration
```
mrregister iw.mif jw.mif -lmax 0,2,4 -scale 0.3,0.5,1 -warp warp.mif
```
going to multi-contrast requires double-dashing existing options that could be contrast-specific which makes it clear that there might be reasons to change option values for each contrast:
```
mrregister iw.mif jw.mif -lmax 0,2,4 -scale 0.3,0.5,1 -warp warp.mif # Fine
mrregister iw.mif jw.mif --contrast ig.mif jg.mif  -lmax 0,2,4 -scale 0.3,0.5,1 -warp warp.mif # ERROR
mrregister iw.mif jw.mif --contrast ig.mif jg.mif  --lmax 0,2,4 -scale 0.3,0.5,1 -warp warp.mif # Fine
```

-----

I wouldn't enforce all multi-argument command to follow any delimiter syntax. unless it makes sense. (I prefer `dwifod dwi.mif rf_w wm.mif rf_g gm.mif -lmax 4,0` over `dwifod dwi.mif :: rf_w wm.mif -lmax 4 :: rf_g gm.mif -lmax 0`). It might make sense to adapt it in some circumstance (for instance for `mrcat`, double-dash for input image-specific options, single dash for global options: 
`mrcat in1.mif in2.mif in3.mif -axis 3 --repeat 1,2,3 -repeat 2` 
would result in an image 
`[in1.mif in2.mif in2.mif in3.mif in3.mif in3.mif in1.mif in2.mif in2.mif in3.mif in3.mif in3.mif]` (#1806)).

**Comment by jdtournier** [comment]
OK, there's a lot of opinions out there... 

Here's my take. I'm reluctant to change the way we handle the command-line parsing in radical ways when simpler solutions exist that will do the job. Most of the options I'm seeing are going to require quite a bit of work to make work, and change the user experience in non-trivial ways. That feels like a big change to introduce, and I'm very reticent to do this unless it is very clearly not just superior, but _necessary_. 

Also, I'm reluctant to introduce syntax that veers too far from what is considered conventional. Not because I don't necessarily think it's a good idea, but because that invariably jars with new users. We want the use of MRtrix apps to be relatively easy to pick up, and ensure we leverage users' experience of using other command-line apps. Bearing in mind many potentially very experienced users may only interact with MRtrix once in a while, and anything out of the ordinary means it'll probably get in the way. 

So on that basis, the more I think about it, the less I like the idea of introducing any form of delimiter in the command. I think it's redundant (we can use the positions on the command-line for the same effect), non-standard (I can't think of any other command that does anything all that similar - even with `foreach` or `startx`, the nature of what comes before and after the delimiter is fundamentally completely different), and I don't find that it necessarily helps clarify the intent. 

If we do decide to introduce delimiters, I'd be very reluctant to use them to delimit arguments to options (e.g. `-lmax 0,2,4 :: 0,2,8`). This would require very deep changes to the parsing backend, since the checks around the number of argument couldn't handle this easily. And as Rob says, I'm not sure it aids with legibility. I'd be much more comfortable with the delimiters being used to group batches of arguments and options, e.g.:
```
command --global --options :: arg1 arg2 --with option :: arg3 arg4 --with other_option
```
or something like that. With the added bonus that you don't have to specify the options for all groups, whereas the in-option grouping forces the user to specify all or nothing, unless you allow things like `-lmax default :: 0,2,8` or something.

But by that point, I really feel the separators are redundant, since the arguments themselves provide unambiguous barrier points. Hence my last suggestion:
```
command --global --options arg1 arg2 --with option arg3 arg4 --with other_option
```
 And I really don't think you'd lose a great deal on the legibility by that point... 

---

I'm also with @dchristiaens on the issue with `dwi2fod`: I don't think it's a problem at all. But more to the point, the idea of simply allowing options to keep track of the argument they come after means that instead of:
```
dwifod dwi.mif rf_w wm.mif rf_g gm.mif -lmax 4,0
```
we could write:
```
dwifod dwi.mif rf_w wm.mif -lmax 4 rf_g gm.mif
```
which strikes me as quite a good solution. We can now set the lmax for each tissue type independently and stick to the default for the others. Feels like a win to me...

---

I'm not a fan of the idea of using single or double dashes for global or per-group options. I don't think it makes it clear at all what is intended, this is such a subtle clue that it's likely to confuse users. It also changes our handling of single or double dashes everywhere - so far, we've accepted as many dashes as users want to throw at the software, mostly because I've always found the distinction pretty artificial and unnecessary. I'd be reluctant to now not only become stricter in what we accept, but on top of that imbue the number of dashes with a meaning that is so far from obvious for a new user...

I'm also not keen on the limitation that you'd be forced to provide the option for all contrasts. The explicit grouping of whole chunks of arguments & options with a delimiter, or simply by their position, both allow options to be set or left to default on a per-group basis without ambiguity. 

---

So I guess what I'm trying to say is that my preference is for the simple positional solution: options that apply per (group of) arguments simply need to be placed directly after them - and before the next (group of) arguments). The reason I prefer this are:

- it leaves the syntax for single contrast completely unchanged. 
- it make very little change to the current command-line parsing 'feel'
- I think it's the closest to convention, and to what you might expect other commands to do (can't think of a good example for this right now though...)
- I _think_ it'll be simpler to implement and use
- it leaves the command-line parsing for all other apps completely unchanged - unless we want them to take advantage of it (e.g. `dwi2fod`). All it means is that apps will have the option of querying which argument comes before each relevant option.
- I think grouping the argument based on their position will be _almost_ as effective in terms of legibility as with explicit delimiters, and certainly equally unambiguous from a technical parsing point of view. 
- it preserves the flexibility of allowing other options to appear in whatever order they want to. For instance, the `-scale` option could be placed at the beginning, at the end, or in the middle, and the app would handle it just the same. 

---

As always, just my 2 cents...

**Comment by Lestropie** [comment]
> My only concern is that additional separation between global options. Ideally, I would prefer the command syntax to remain unchanged for single-contrast registration.

I tried to be careful with my wording on 2a, but probably should have been more explicit:

> - Options that apply globally and hence must appear before the first colon (*if any are present*);

> - Options that cannot appear before the first colon (*if present*) as they apply only to "argument groups".

By the precise wording of this logic, if there is only one contrast present, no delimiter is necessary. Conversely it is also acceptable to include the delimiter even if only one contrast is present: all that does is introduce enforcement of which options appear before vs. after the delimiter.

-----

> `-option option_args` is for global options
> `--option option_args` is for contrast-related options, this is made clear in the command help page

Not sold on it given precedence. E.g. many Unix commands have the same option with two different usages: a longer string with "`--`" or a compact usage with "`-`". 

-----

> (I prefer `dwifod dwi.mif rf_w wm.mif rf_g gm.mif -lmax 4,0` over `dwifod dwi.mif :: rf_w wm.mif -lmax 4 :: rf_g gm.mif -lmax 0`).

To clarify, I'm not particularly sold on that change myself: I simply wanted to ensure that all possible responses to the statement "`mrregister` usage no longer matches `dwi2fod`" are explored. This is a feasible response to such; another is prioritising the interface of `mrregister` over how it may or may not conflict with `dwi2fod`.

-----

> I guess what I'm trying to say is that my preference is for the simple positional solution: options that apply per (group of) arguments simply need to be placed directly after them - and before the next (group of) arguments).

I agree that this is pretty easy to implement in the C++ back-end, probably gives all of the capability needed, and is less likely to be alienating. Note however that it may not be possible in Python without a complete re-write of argument parsing (#1608).

Another thing to consider with this type of approach is that if explicit delimiters are not present, one could still use newlines ("`\`") in the terminal / a shell script file to provide explicit grouping of contrasts and their relevant options.

**Comment by maxpietsch** [comment]
I think the most common command-line usage is close to the one defined in the [posix standard](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01
) that uses options (optionally with corresponding arguments), in any order, followed by arguments. Single-dash is used for flags (single-letter options), double-dash for spelt out options, and double-dash by itself (not as the start of an option) indicates the end of options and start of arguments.

I like about the MRtrix argument parsing that it is flexible in where options appear (before, in the middle, or after arguments), that options can be abbreviated if possible, and that the order of options does not matter unless they interact. -- Tentative statement of my opinion.

---


> Here's my take. I'm reluctant to change the way we handle the command-line parsing in radical ways when simpler solutions exist that will do the job. Most of the options I'm seeing are going to require quite a bit of work to make work, and change the user experience in non-trivial ways. That feels like a big change to introduce, and I'm very reticent to do this unless it is very clearly not just superior, but _necessary_.

100% agree

> Also, I'm reluctant to introduce syntax that veers too far from what is considered conventional. Not because I don't necessarily think it's a good idea, but because that invariably jars with new users. We want the use of MRtrix apps to be relatively easy to pick up, and ensure we leverage users' experience of using other command-line apps. Bearing in mind many potentially very experienced users may only interact with MRtrix once in a while, and anything out of the ordinary means it'll probably get in the way.

100% agree on maximising usability unless it is a large developmental burden.

> So on that basis, the more I think about it, the less I like the idea of introducing any form of delimiter in the command. I think it's redundant (we can use the positions on the command-line for the same effect), non-standard (I can't think of any other command that does anything all that similar - even with foreach or startx, the nature of what comes before and after the delimiter is fundamentally completely different), and I don't find that it necessarily helps clarify the intent.

I don't particularly like having a delimiter either. However, I prefer using it over having ambiguity in what options refer to (case 1b). 

> If we do decide to introduce delimiters, I'd be very reluctant to use them to delimit arguments to options (e.g. -lmax 0,2,4 :: 0,2,8). This would require very deep changes to the parsing backend, since the checks around the number of argument couldn't handle this easily.

I think Rob's point on legibility is a valid one `command arg1 :: arg2 -option option_arg1 :: option_arg2` is not very readable. You know the code far better but I just don't see a problem of implementing this as a translation of the arguments to ``command arg1 arg2 -option option_arg1 -option option_arg2`. This preprocessing step (find `::`, check if the previous argument was part of an option (is this where it breaks?), if so replace the `::` with `-option` else ignore) is literally all that is necessary for this to work. Anyway, not my preferred solution.

> I really feel the separators are redundant, since the arguments themselves provide unambiguous barrier points. ... 
> `command --global --options arg1 arg2 --with option arg3 arg4 --with other_option`
> And I really don't think you'd lose a great deal on the legibility by that point...

I don't I agree on this one. I find this confusing and not consistent with existing MRtrix or POSIX command-line parsing standards and feels restricting. To parse this the user needs to know

a) that options are grouped by arguments in that particular command (they aren't for any other command so far) 
b) that grouping either applies only to non-global options or that global options need to come first 
c) how defaults are used: are they overwritten for all arguments after the one where it's set or is the default used unless specified

All of this could be signalled by a double dash without the need to adhere to any particular order other than that of repeated options.

`command -global -options arg1 arg2 --args arg3 arg4 --with option --with other_option`
or equivalently
`command -global -options arg1 arg2 --with option --args arg3 arg4 --with other_option`

>It also changes our handling of single or double dashes everywhere - so far, we've accepted as many dashes as users want to throw at the software
I didn't know about this until recently. I doubt many people use anything but what we documented in the command help pages. (There is always `-option` vs `+option` vs`:option`...)

> I'd be reluctant to now not only become stricter in what we accept ...

I think the single and double dash solution is far less restrictive than grouping by arguments but allowing arbitrary number of dashes.

> but on top of that imbue the number of dashes with a meaning that is so far from obvious for a new user...
Admittedly, the double dash options are also not POSIX style compliant
`command -global -options --with option --with other_option arg1 arg2 --args arg3 arg4 `
but feel closer than grouping by argument and option globality.

Let's assume you're new to `dwi2fod`. In `dwifod dwi.mif rf_w wm.mif -lmax 4 rf_g gm.mif`
how would you know that `-lmax` applies only to the first argument pair and that it's not just another command where the order of options does not matter?

I wouldn't change `dwi2fod' but if I did, I'd prefer 
`dwi2fod -grad grad dwi.mif rf_w wm.mif --lmax 4 rf_g gm.mif` 
being equal to 
`dwi2fod dwi.mif rf_w wm.mif rf_g gm.mif --lmax 4 -grad grad `
if possible.

As you said, a double dash is a subtle hint but it is a hint that this parameter is special and if omitted where necessary, the error message could state this.

> I'm also not keen on the limitation that you'd be forced to provide the option for all contrasts.

For one or all but as you wrote, a `default` option argument could work as well. More verbose but clear:
`command arg1 arg2 arg3 --more barg1 barg2 barg3 --opt 1 --opt default --opt 3 -faster`

----
Double-dash (or plus-prefixed) options if needed:

- it leaves the syntax for single contrast completely unchanged. ✓
- it make very little change to the current command-line parsing 'feel' ✓
- I think it's the closest to convention, and to what you might expect other commands to do (can't think of a good example for this right now though...) ✓
- I _think_ it'll be simpler to implement and use ✓
- it leaves the command-line parsing for all other apps completely unchanged - unless we want them to take advantage of it (e.g. `dwi2fod`). ✓
- allows grouping by argument if the user chooses to but is not necessary, equally unambiguous as current command line parsing from a technical parsing point of view. ✓
- it preserves the flexibility of allowing other options to appear in whatever order they want to. For instance, the `-scale` option could be placed at the beginning, at the end, or in the middle, and the app would handle it just the same. ✓
- is explicit in what global vs grouped-specific options are unless not necessary ✓
